internal/controller/apiextensions/claim/syncer_csa.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package claim
+
+import (
+	"context"
+	"fmt"
+
+	"dario.cat/mergo"
+	"github.com/google/go-cmp/cmp"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/claim"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
+
+	"github.com/crossplane/crossplane/internal/names"
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+const (
+	errUpdateClaim          = "cannot update claim"
+	errUnsupportedClaimSpec = "claim spec was not an object"
+	errGenerateName         = "cannot generate a name for composite resource"
+	errApplyComposite       = "cannot apply composite resource"
+
+	errMergeClaimSpec   = "unable to merge claim spec"
+	errMergeClaimStatus = "unable to merge claim status"
+)
+
+// A ClientSideCompositeSyncer binds and syncs a claim with a composite resource
+// (XR). It uses client-side apply to update the claim and the composite.
+type ClientSideCompositeSyncer struct {
+	client resource.ClientApplicator
+	names  names.NameGenerator
+}
+
+// NewClientSideCompositeSyncer returns a CompositeSyncer that uses client-side
+// apply to sync a claim with a composite resource.
+func NewClientSideCompositeSyncer(c client.Client, ng names.NameGenerator) *ClientSideCompositeSyncer {
+	return &ClientSideCompositeSyncer{
+		client: resource.ClientApplicator{
+			Client:     c,
+			Applicator: resource.NewAPIPatchingApplicator(c),
+		},
+		names: ng,
+	}
+}
+
+// Sync the supplied claim with the supplied composite resource (XR). Syncing
+// may involve creating and binding the XR.
+func (s *ClientSideCompositeSyncer) Sync(ctx context.Context, cm *claim.Unstructured, xr *composite.Unstructured) error { //nolint:gocyclo // This complex process seems easier to follow in one long method.
+	// First we sync claim -> XR.
+
+	// It's possible we're being asked to configure a statically provisioned XR.
+	// We should respect its existing external name annotation.
```

Comment:
Looks like there was a Non-Breaking Space:
```suggestion
	// We should respect its existing external name annotation.
```